contrib/checkpatch/checkpatch.pl

```diff
@@ -0,0 +1,6937 @@
+#!/usr/bin/env perl
+# SPDX-License-Identifier: GPL-2.0
+#
+# (c) 2001, Dave Jones. (the file handling bit)
+# (c) 2005, Joel Schopp <jschopp@austin.ibm.com> (the ugly bit)
+# (c) 2007,2008, Andy Whitcroft <apw@uk.ibm.com> (new conditions, test suite)
+# (c) 2008-2010 Andy Whitcroft <apw@canonical.com>
+# (c) 2010-2018 Joe Perches <joe@perches.com>
+
+use strict;
+use warnings;
+use POSIX;
+use File::Basename;
+use Cwd 'abs_path';
+use Term::ANSIColor qw(:constants);
+use Encode qw(decode encode);
+
+my $P = $0;
+my $D = dirname(abs_path($P));
+
+my $V = '0.32';
+
+use Getopt::Long qw(:config no_auto_abbrev);
+
+my $quiet = 0;
+my $tree = 1;
+my $chk_signoff = 1;
+my $chk_patch = 1;
+my $tst_only;
+my $emacs = 0;
+my $terse = 0;
+my $showfile = 0;
+my $file = 0;
+my $git = 0;
+my %git_commits = ();
+my $check = 0;
+my $check_orig = 0;
+my $summary = 1;
+my $mailback = 0;
+my $summary_file = 0;
+my $show_types = 0;
+my $list_types = 0;
+my $fix = 0;
+my $fix_inplace = 0;
+my $root;
+my %debug;
+my %camelcase = ();
+my %use_type = ();
+my @use = ();
+my %ignore_type = ();
+my @ignore = ();
+my $help = 0;
+my $configuration_file = ".checkpatch.conf";
+my $max_line_length = 100;
+my $ignore_perl_version = 0;
+my $minimum_perl_version = 5.10.0;
+my $min_conf_desc_length = 4;
+my $spelling_file = "$D/spelling.txt";
+my $codespell = 0;
+my $codespellfile = "/usr/share/codespell/dictionary.txt";
+my $conststructsfile = "$D/const_structs.checkpatch";
```

Comment:
Couldn't we remove references to this file in the script instead of having an empty version? Unless you plan to keep this script up-to-date with the kernel's and want to avoid diverging too much? If that's the case, maybe it would be easier to download it directly from the kernel sources in the GitHub action.

Reply:
Updating the script from the kernel repository from time to time was my intention indeed, so I'd rather avoid patching it if we don't have a strong incentive to do so. Having one empty file under contrib/checkpatch does not seem too expensive?

I thought of downloading the script from the kernel version, but I'd rather keep control of the version we use to avoid bad surprise (e.g. some change regarding a new file needed, that breaks checkpatch if the file is not found and makes all runs of the GitHub action fail all of a sudden.

Regarding the empty file, instead of patching our checkpatch version, I was considering maybe patching upstream so that it doesn't complain if the file is missing but `--ignore CONST_STRUCT` is passed. What do you think? (This specific check is also broken if `CONST_STRUCT` is not ignored but the file is present and empty by the way, so even if we don't really care in our case I could fix that too).

Reply:
> Regarding the empty file, instead of patching our checkpatch version, I was considering maybe patching upstream so that it doesn't complain if the file is missing but `--ignore CONST_STRUCT` is passed. What do you think? (This specific check is also broken if `CONST_STRUCT` is not ignored but the file is present and empty by the way, so even if we don't really care in our case I could fix that too).

:+1: if you have the time; otherwise LGTM!

Reply:
It doesn't look too hard (even though it's Perl!). I'll post that soon.

<details><summary>patch</summary>

``` diff
diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index bf9e0e87a6ef..d91332360494 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -765,8 +765,10 @@ sub read_words {
 }
 
 my $const_structs = "";
-read_words(\$const_structs, $conststructsfile)
-    or warn "No structs that should be const will be found - file '$conststructsfile': $!\n";
+if (show_type("CONST_STRUCT")) {
+       read_words(\$const_structs, $conststructsfile)
+           or warn "No structs that should be const will be found - file '$conststructsfile': $!\n";
+}
 
 my $typeOtherTypedefs = "";
 if (length($typedefsfile)) {
@@ -6580,7 +6582,8 @@ sub process {
 
 # check for various structs that are normally const (ops, kgdb, device_tree)
 # and avoid what seem like struct definitions 'struct foo {'
-               if ($line !~ /\bconst\b/ &&
+               if ($const_structs ne "" &&
+                   $line !~ /\bconst\b/ &&
                    $line =~ /\bstruct\s+($const_structs)\b(?!\s*\{)/) {
                        WARN("CONST_STRUCT",
                             "struct $1 should normally be const\n" . $herecurr);
```
</details>